batcher/aligned-sdk/src/communication/messaging.rs

```diff
@@ -0,0 +1,213 @@
+            if let Some(close_msg) = close_frame {
+                error!("Connection was closed before receiving all messages. Reason: {}. Try submitting your proof again", close_msg.to_owned());
+                ws_write.lock().await.close().await?;
+                return Ok(None);
+            }
+            error!("Connection was closed before receiving all messages. Try submitting your proof again");
+            ws_write.lock().await.close().await?;
+            return Ok(None);
+        } else {
+            process_batch_inclusion_data_without_await(
+                msg,
+                &mut aligned_verification_data,
+                verification_data_commitments_rev,
+                num_responses.clone(),
+            )
+            .await?;
+
+            if *num_responses.lock().await == total_messages {
+                debug!("All messages responded. Closing connection...");
+                ws_write.lock().await.close().await?;
+                return Ok(Some(aligned_verification_data));
+            }
+        }
+    }
+
+    Ok(None)
+}
+
+pub async fn receive_and_wait(
+    ws_read: SplitStream<WebSocketStream<MaybeTlsStream<TcpStream>>>,
+    ws_write: Arc<Mutex<SplitSink<WebSocketStream<MaybeTlsStream<TcpStream>>, Message>>>,
+    eth_rpc_provider: Provider<Http>,
+    contract_address: &str,
+    total_messages: usize,
+    num_responses: Arc<Mutex<usize>>,
+    verification_data_commitments_rev: &mut Vec<VerificationDataCommitment>,
+) -> Result<Option<Vec<AlignedVerificationData>>, SubmitError> {
+    // Responses are filtered to only admit binary or close messages.
+    let mut response_stream =
+        ws_read.try_filter(|msg| future::ready(msg.is_binary() || msg.is_close()));
+
+    let mut aligned_verification_data: Vec<AlignedVerificationData> = Vec::new();
+
+    let service_manager = aligned_service_manager(eth_rpc_provider.clone(), contract_address)
+        .await
+        .map_err(|e| SubmitError::EthError(e.to_string()))?;
+
+    let events = service_manager.event::<BatchVerifiedFilter>();
+
+    let mut stream = events
+        .stream()
+        .await
+        .map_err(|e| SubmitError::BatchVerifiedEventStreamError(e.to_string()))?;
+
+    // Two different proofs can return the same batch_merkle_root and we don't want wait for the same event twice.
+    let mut verified_batch_merkle_roots = HashSet::new();
+
+    while let Some(Ok(msg)) = response_stream.next().await {
+        if let Message::Close(close_frame) = msg {
+            if let Some(close_msg) = close_frame {
+                error!("Connection was closed before receiving all messages. Reason: {}. Try submitting your proof again", close_msg.to_owned());
+                ws_write.lock().await.close().await?;
+                return Ok(None);
+            }
+            error!("Connection was closed before receiving all messages. Try submitting your proof again");
+            ws_write.lock().await.close().await?;
+            return Ok(None);
```

Comment:
same comment about returning error

Reply:
Solved in [cef4afb](https://github.com/yetanotherco/aligned_layer/pull/583/commits/cef4afb1e728da34f45728b42d6780e29a862585)